ext/nnstreamer/android_source/gstamcsrc.c

```diff
@@ -0,0 +1,987 @@
+      /** No output buffer right now */
+    } else {
+      LOGE("Unexpected info code: %zd", buf_idx);
+    }
+  }
+
+  /** unless it reaches to EOS, check buffers again */
+  if (!priv->sawInputEOS || !priv->sawOutputEOS)
+    Looper_post (priv->looper, MSG_CODEC_BUFFER, self, FALSE); 
+}
+
+/**
+ * @brief looper internal function to handle each command
+ */
+static void looper_handle (gint cmd, void *data)
+{
+  if (data != NULL) {
+    GstAMCSrc *self = GST_AMC_SRC (data);
+    GstAMCSrcPrivate *priv = GST_AMC_SRC_GET_PRIVATE (self);
+
+    switch (cmd) {
+      case MSG_CODEC_BUFFER:
+        /** main codec function */
+        check_codec_buf (self);
+        break;
+      case MSG_CODEC_DONE:
+        /** stop media codec */
+        AMediaCodec_stop (priv->codec);
+        AMediaCodec_delete (priv->codec);
+        AMediaExtractor_delete (priv->ex);
+
+        priv->sawInputEOS = TRUE;
+        priv->sawOutputEOS = TRUE;
+
+        LOGI ("Finished");
+        break;
+      case MSG_CODEC_SEEK:
+        AMediaExtractor_seekTo (priv->ex, 0, AMEDIAEXTRACTOR_SEEK_NEXT_SYNC);
+        AMediaCodec_flush (priv->ex);
+
+        priv->renderstart = -1;
+        priv->sawInputEOS = FALSE;
+        priv->sawOutputEOS = FALSE;
+
+        if (!priv->isPlaying) {
+          priv->renderonce = TRUE;
+          Looper_post (priv->looper, MSG_CODEC_BUFFER, self, FALSE);
+        }
+
+        LOGI ("Seeked");
+        break;
+      case MSG_CODEC_PAUSE:
+        if (priv->isPlaying) {
+          priv->isPlaying = FALSE;
+
+          Looper_post (priv->looper, MSG_CODEC_PAUSE_ACK, self, TRUE);
+
+          LOGI ("Paused");
+        }
+        break;
+      case MSG_CODEC_PAUSE_ACK:
+        /** Only used to flush buffers */
+        break;
+      case MSG_CODEC_RESUME:
+        if (!priv->isPlaying) {
+          priv->renderstart = -1;
+          priv->isPlaying = TRUE;
+
+          Looper_post (priv->looper, MSG_CODEC_BUFFER, self, FALSE);
+
+          LOGI ("Resumed");
+        }
+        break;
+      default:
+        LOGE ("Unexpected message: cmd(%d)", cmd);
+        break;
+    }
+  }
+}
+
+/**
+ * @brief start function, called when state changed null to ready
+ * start the media codec and its looper
+ */
+static gboolean
+gst_amc_src_start (GstBaseSrc * src)
+{
+  GstAMCSrc *self = GST_AMC_SRC (src);
+  GstAMCSrcPrivate *priv = GST_AMC_SRC_GET_PRIVATE (self);
+
+  /** setup a looper */
+  priv->looper = Looper_new();
+  Looper_set_handle (priv->looper, looper_handle);
+
+  priv->previous_ts = GST_CLOCK_TIME_NONE;
+  priv->started = TRUE;
+
+  /** start media codec */
+  AMediaCodec_start(priv->codec);
+  Looper_post (priv->looper, MSG_CODEC_BUFFER, self, false);
+
+  return TRUE;
+}
+
+/**
+ * @brief stop function, called when state changed ready to null
+ */
+static gboolean gst_amc_src_stop (GstBaseSrc * src)
+{
+  GstAMCSrc *self = GST_AMC_SRC (src);
+  GstAMCSrcPrivate *priv = GST_AMC_SRC_GET_PRIVATE (self);
+
+  g_mutex_lock (&priv->mutex);
+
```

Comment:
reset `priv->previous_ts` here? Otherwise duration might be spurious when element restarts, without de-allocation.

Reply:
Yeap, you're right. It'll be fixed.